quickwit/quickwit-ingest/src/ingest_v2/workbench.rs

```diff
@@ -48,22 +54,77 @@ impl IngestWorkbench {
 
         Self {
             subworkbenches,
-            num_successes: 0,
+            max_num_attempts,
+            ..Default::default()
         }
     }
 
+    pub fn new_attempt(&mut self) {
+        self.num_attempts += 1;
+    }
+
+    pub fn is_complete(&self) -> bool {
+        self.num_successes >= self.subworkbenches.len()
+            || self.num_attempts >= self.max_num_attempts
+            || self.has_no_pending_subrequests()
+    }
+
+    pub fn is_last_attempt(&self) -> bool {
+        self.num_attempts >= self.max_num_attempts
+    }
+
+    fn has_no_pending_subrequests(&self) -> bool {
+        self.subworkbenches
+            .values()
+            .all(|subworbench| !subworbench.is_pending())
+    }
+
     #[cfg(not(test))]
     pub fn pending_subrequests(&self) -> impl Iterator<Item = &IngestSubrequest> {
         self.subworkbenches.values().filter_map(|subworbench| {
-            if !subworbench.is_success() {
+            if subworbench.is_pending() {
                 Some(&subworbench.subrequest)
             } else {
                 None
             }
         })
     }
 
-    pub fn record_success(&mut self, persist_success: PersistSuccess) {
+    pub fn record_get_or_create_open_shards_failure(
+        &mut self,
+        open_shards_failure: GetOrCreateOpenShardsFailure,
+    ) {
+        let Some(subworkbench) = self
+            .subworkbenches
+            .get_mut(&open_shards_failure.subrequest_id)
+        else {
+            warn!(
+                "could not find subrequest `{}` in workbench",
+                open_shards_failure.subrequest_id
+            );
+            return;
+        };
+        subworkbench.num_attempts += 1;
+
+        let last_failure_opt = match open_shards_failure.reason() {
+            GetOrCreateOpenShardsFailureReason::IndexNotFound => {
+                Some(SubworkbenchFailure::IndexNotFound)
+            }
+            GetOrCreateOpenShardsFailureReason::SourceNotFound => {
+                Some(SubworkbenchFailure::SourceNotFound)
+            }
+            GetOrCreateOpenShardsFailureReason::Unspecified => {
+                warn!(
+                    "failure reason for subrequest `{}` is unspecified",
+                    open_shards_failure.subrequest_id
+                );
+                None
```

Comment:
I don't understand why we record `None` here. 
We seem to have logic to different retryable (you called that transient) from non retryable errors.

Could we reserve `None` to we have never received a failure yet?